### PR TITLE
Added Video Morph

### PR DESCRIPTION
**Pull request by rickmcgeer**
This isn't quite ready for prime time, But I would appreciate feedback.  The big issue is with initialization -- it works fine as long as I don't specify properties to the constructor, but when I do it hangs.
```
new VideoMorph().openInWorld()
```
works fine, but
```
new VideoMorph().openInWorld({loop: true})
```
hangs.
```
$world.execCommand("open browser", {moduleName: "html-morph.js", packageName: "lively.morphic", codeEntity: "VideoMorph"});
```
I should revise before merge.
There are other usability questions (e.g., adding a menu for property settings, and should the HMLT be editable?

### CODE REVIEW

**Comment by linusha** [comment]
@rickmcgeer I will take a more in-depth look tomorrow, but it could be that the problem with the initialization has nothing per se to do with your code and is instead a bug in the system. I will report back to you!

Regarding the HTML, imo it should not be editable.
In general, while I see the benefit that a menu etc could provide from a UX perspective, I tend to think that the more stuff like this is wanted, the more one would benefit from just itegrating an already established JS video player and wrapping that in a morph. Those come with all kinds of UI elements already and mimic what people are used to.

File: lively.morphic/html-morph.js [hunk 1]
```diff
@@ -271,3 +271,196 @@ export class IFrameMorph extends HTMLMorph {
...
+  }
+
+  onLoad () {
+    const props = this.constructorProps;
+    const setField = (field, defaultVal) => {
+      this[field] = props && props[field] ? props[field] : defaultVal;
+    };
+    const normalizeBoolean = field => {
+      this[field] = props ? !!props[field] : false;
+    };
```
[the remark this answers is not in the document]
Comment (rickmcgeer) [reply lively.morphic/html-morph.js:366]: Well, the values really aren't problematic, I guess, because JavaScript treats any value other than `0`, `undefined`, `null`, and `false` as `false`.  I don't really know how having a non-Boolean value in a Boolean property could hurt, but for display in inspectors, etc., I usually try to normalize them.

File: lively.morphic/html-morph.js [hunk 2]
```diff
@@ -271,3 +271,196 @@ export class IFrameMorph extends HTMLMorph {
...
+    fields.forEach(field => setField(field.name, field.defaultVal));
+    this.resetHTML();
+  }
+
+  resizeToExtentProperties () {
+    this.whenRendered().then(_ => {
+      if (this.autosize) {
+        this.resizeToNaturalExtent();
+      } else if (this.autoaspect) {
+        this.resizeToFitVideo();
```
[the remark this answers is not in the document]
Comment (rickmcgeer) [reply lively.morphic/html-morph.js:389]: This is a great idea!

File: lively.morphic/html-morph.js [hunk 3]
```diff
@@ -271,3 +271,196 @@ export class IFrameMorph extends HTMLMorph {
...
+          this.setProperty('autosize', !!aBool);
+          this.resizeToExtentProperties();
+        }
+      }
+
+    };
+  }
+
+  constructor (props) {
+    this.videoId = `video-${Math.floor(Math.random() * 1E6)}`;
```
[the remark this answers is not in the document]
Comment (rickmcgeer) [reply lively.morphic/html-morph.js:349]: This is an artifact from when I was searching for the DOM element using `document.getElementById()`.  This is a common problem that I have with Lively, when  an API requires the DOM element.  I haven't found a reasonable way to get the DOM element from Lively in a reliable way.  In this case, the DOM element is required to control play, stop, rewind, etc.
So this should come out, if you can suggest reasonable code to get the DOM element.

**Comment by rickmcgeer** [comment]
@linusha thanks for the feedback! Regarding other libraries, the reason I used the HTML5 native libraries was (a) familiarity; and (b) a desire to avoid code bloat -- downloading and starting Lively is already pretty daunting, just due to the sheer volume of npm libraries we use.  But we can use something else, if you have suggestions.  @merryman should probably also weigh in.
I've been using the HTML edit feature primarily for debugging and looking at the generated HTML.  We could make this read-only for that purpose, or simply dump it -- it just involves overriding `menuItems()` and removing the items that `HTMLMorph.menuItems()` adds.

**Comment by rickmcgeer** [comment]
HTML/CSS are no longer editable as of the latest commit (332dfb39b76e86c0096a5d4737aa8e990d36a9b2)

File: lively.morphic/html-morph.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (linusha) [reply lively.morphic/html-morph.js:366]: Got it, makes sense!

**Comment by linusha** [comment]
Alright,

1. The initialization code you have above plugs the arguments into the wrong method, `{loop: true}` needs to be passed into the `VideoMorph` constructor, not into `openInWorld()`.
2. I think for now it is okay to stick with just managing the native `video` element manually. It depends heavily on the desired complexity/feature set. When at some point e.g. a timeline with preview hovering or something like that is needed, it would not make much sense to reimplement that.
3. Adding a library causes additional complexity, agreed. And as I said, as long as the feature set we need is basically what the native `video` element provides anyways I agree that it is the correct decision to avoid that complexity. However, in case those features were necessary, bundle size should not be the reason for reinventing the wheel. If that were the limiting factor, we need to revisit the bundle generation and other quirks that might make for bad UX when working with dependencies.

File: lively.morphic/html-morph.js [hunk 4]
```diff
@@ -271,3 +271,205 @@ export class IFrameMorph extends HTMLMorph {
...
+    bools.forEach(bool => normalizeBoolean(bool));
+    const fields = [
+      { name: 'src', defaultVal: 'https://matt.engagelively.com/assets/kaleidoscope-art-17141.mp4' },
+      { name: 'type', defaultVal: 'video/mp4' },
+      { name: 'badBrowserMessage', defaultVal: 'Your browser does not support video content' },
+      { name: 'loop', defaultVal: true },
+      { name: 'autoplay', defaultVal: true },
+      { name: 'controls', defaultVal: false },
+      { name: 'autosize', defaultVal: false },
+      { name: 'autoaspect', defaultVal: true }
```
Comment (linusha) [inline lively.morphic/html-morph.js:378]: Can we make the defaultValues here consistent with the ones specified for the morphic properties above?

Comment (rickmcgeer) [reply lively.morphic/html-morph.js:378]: Sure

File: lively.morphic/html-morph.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rickmcgeer) [reply lively.morphic/html-morph.js]: @linusha this has now been replaced with a single property, `videoLayout` with three possible values `autosize`, `autoaspect` and `none`.

File: lively.morphic/html-morph.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rickmcgeer) [reply lively.morphic/html-morph.js]: Now removed.

File: lively.morphic/html-morph.js (hunk 4 above)
Comment (rickmcgeer) [reply lively.morphic/html-morph.js]: This is now done.

**Comment by rickmcgeer** [comment]
I have now addressed the issues above, and thanks @linusha for everything.
One thing is that I have turned off `autosize` and `autoaspect` on initialization, because there were (of course) timing issues.  The issue is that  these are infeasible until the video is loaded, because the native extent of the video isn't known until load.

**Comment by linusha** [comment]
Hi @rickmcgeer,

thanks for taking care of the points I raised!
I just took the liberty of cleaning some things up, mainly the following:

- rewrote history to adhere to the lively.next repo conventions
- removed commented out code (mainly console.logs)
- added a `pausePlaying` method
- made sure that the video stops playing when the morph is removed
- fixed the `menuItems` to be resilient against changes in the `menuItems` of the `HTMLMorph`
- extracted the whole thing into its own module in `lively.components` (see #807).

From our side, this is thus good to go. If there is nothing that you want to add to this feature right now, please give me a quick heads up here and then I will merge this! 
Thanks for your work!

**Comment by rickmcgeer** [comment]
@linusha I found a timing bug in frozen code on some browsers, due to nulls in either bounds or the `videoDomElement`, so I caught the error and checked it in.  Can you please check to make sure I didn't walk over your changes?  I did a pull before I modified the code, and got an up-to-date, but I noticed there was still some commented-out code.

**Comment by linusha** [comment]
> @linusha I found a timing bug in frozen code on some browsers, due to nulls in either bounds or the `videoDomElement`, so I caught the error and checked it in. Can you please check to make sure I didn't walk over your changes? I did a pull before I modified the code, and got an up-to-date, but I noticed there was still some commented-out code. And as I look at the repo I don't see the changes you mentioned. Did you check them in? In any case, the description of your changes look good to me. I can look at the code if you like, but I'm sure it's done well.

Yes, I checked them in but as I said, I moved the `VideoMorph` into its own module inside of `lively.component`. I also removed it from the `html-morph.js` module completely. What happened here is the following: I force pushed to clean up the git history, making it impossible for git to just forward your branch to the version of the remote. Therefore, git created a merge commit so that your changes and my cleaned up version in the other module existed side-by-side and then you just added code into your version.

I (and @merryman in the meantime as well) use force-push rather liberally to keep the history nice and tidy, which means you will either need to reset your local HEAD **when you do not have local changes** (`git reset --hard origin/BRANCH-NAME` or rebase your local changes on the remote (`git pull --rebase origin BRANCHNAME`). This does not guarantee a correct result, but should work most of the times.

If you have problems with this git stuff always feel free to ping me. As a rule of thumb, if you get a merge commit, something is wrong, as we do not use them in lively.next but instead rely on a rebase-driven approach.

TLDR: I took care of it, all changes are now incorporated. **Note, that the module is now `lively.components/video.js`**.
I will merge this for now :tada: to reduce the potential for further annoyances with git, if you find other problems, just feel free to open another PR!

**Review by linusha — APPROVED** [review]
